OpenRA.Mods.Common/Traits/BotModules/UnitBuilderBotModule.cs

```diff
@@ -161,32 +147,39 @@ void BuildUnit(IBot bot, string name)
 
 		ActorInfo ChooseRandomUnitToBuild(ProductionQueue queue)
 		{
-			var buildableThings = queue.BuildableItems();
-			if (!buildableThings.Any())
+			var unitsToBuild = Info.UnitsToBuild.ToArray();
+			if (unitsToBuild.Length == 0)
 				return null;
 
-			var unit = buildableThings.Random(world.LocalRandom);
-			return HasAdequateAirUnitReloadBuildings(unit) ? unit : null;
-		}
-
-		ActorInfo ChooseUnitToBuild(ProductionQueue queue)
-		{
-			var buildableThings = queue.BuildableItems();
-			if (!buildableThings.Any())
+			var buildableThings = queue.BuildableItems().Shuffle(world.LocalRandom).ToArray();
+			if (buildableThings.Length == 0)
```

Comment:
Do this check first, before shuffling and empty list and creating an array out of an empty list.

Reply:
`queue.BuildableItems()` returns `IEnumerable` and if you do this check by using `Count()` then it will convert to a collection. If there is too many items, check the list first will cause one more unnecessary convertion on the worse case. Meanwhile, when `BuildableItems()` has nothing, the convertion of this line will be very cheap.

Reply:
I don't know if `Any()` will also result in a conversion, considering `BuildableItems` returns a non-empty `KeyCollection` on the worst case. I think we'd better avoid Linq here.